File: R/analyze_mfi.R
## Graphs MFI for each cell population
## There are three different ways of calculating MFI
## 1. Geometric mean: This is the best choice, because the calculation compensates
## for extreme skew, which is standard in flow cytometry data due to the logarithmic
## nature of flow cytometry data.
## 2. Median is a good second choice, however, in situations where the distribution
## is an exponential decay with 0 being the most common value, the median will not
## accurately represent the distribution
## 3. The Arithmetic Mean (ie. the Mean) is a bad choice. It will systemically
## overestimate the representative value due to the logarithmic nature of flow
## cytometry data.
## This script treats all three kinds of MFI equally, however, you can rename it on
## graph using the -y option

wd = dirname(this.path::here())  # wd = '~/github/R/helperFlowCell'
suppressPackageStartupMessages(library('dplyr'))
suppressPackageStartupMessages(library('logr'))
library('optparse')
import::from(progress, 'progress_bar')
import::from(jsonlite, 'toJSON')
import::from(stringr, 'str_detect')
import::from(tidyr, 'pivot_longer')
import::from(ggplot2, 'ggsave')

import::from(file.path(wd, 'R', 'functions', 'preprocessing.R'),
    'sort_groups_by_metric', .character_only=TRUE)
import::from(file.path(wd, 'R', 'functions', 'file_io.R'),
    'import_flowjo_export', 'import_flow_metadata', .character_only=TRUE)

import::from(file.path(wd, 'R', 'tools', 'df_tools.R'),
    'group_by_agg', 'rename_columns',
    .character_only=TRUE)
import::from(file.path(wd, 'R', 'tools', 'file_io.R'),
    'append_many_csv', .character_only=TRUE)
import::from(file.path(wd, 'R', 'tools', 'list_tools.R'),
    'items_in_a_not_b', 'multiple_replacement', 'move_list_items_to_front',
    .character_only=TRUE)
import::from(file.path(wd, 'R', 'tools', 'math.R'),
    'apply_unpaired_t_test', 'apply_multiple_comparisons',
    'generate_lognormal_data', 'compute_normal_tvd',
    .character_only=TRUE)
import::from(file.path(wd, 'R', 'tools', 'plotting.R'),
    'save_fig', 'plot_violin', 'plot_multiple_comparisons',
    'plot_modal_histograms', .character_only=TRUE)
import::from(file.path(wd, 'R', 'config', 'flow.R'),
    'mouse_db_ignore', .character_only=TRUE)
import::from(file.path(wd, 'R', 'config', 'user_input.R'),
    'custom_group_order', .character_only=TRUE)


# ----------------------------------------------------------------------
# Pre-script settings

# args
option_list = list(
    make_option(c("-i", "--input-dir"), default='data/flow-mfi',
                metavar='data/flow-gmfi', type="character",
                help="input directory, all csv files will be read in"),

    make_option(c("-o", "--output-dir"), default="hfc-output",
                metavar="hfc-output", type="character",
                help="output directory for the data"),

    make_option(c("-d", "--sdev-dir"), default='data/flow-sdev',
                metavar='data/flow-sdev', type="character",
                help="standard deviations"),

    make_option(c("-c", "--counts-dir"), default='data/flow-counts',
                metavar='data/flow-counts', type="character",
                help="counts"),

    make_option(c("-m", "--metadata-dir"), default="data/flow-metadata",
                metavar="data/flow-metadata", type="character",
                help="metadata directory containing csvs mapping fcs to mouse ids"),

    make_option(c("-r", "--ref-dir"), default='data/mice',
                metavar='data/mice', type="character",
                help="directory of files containing all the mouse data"),

    make_option(c("-g", "--group-by"), default='sex,treatment,zygosity',
                metavar='treatment', type="character",
                help="enter a column or comma-separated list of columns, no spaces"),

    make_option(c("-p", "--plot-histograms"), default=FALSE, action="store_true",
                metavar="FALSE", type="logical",
                help="generate histograms"),

    make_option(c("-y", "--metric"), default="gmfi",
                metavar="gmfi", type="character",
                help="metric name, eg. opt[['metric']] or 'mode'"),

    make_option(c("-s", "--stat"), default='fishers_lsd',
                metavar='fishers_lsd', type="character",
                help="Choose 'fishers_lsd', 't_test', 'tukey', or 'bonferroni'"),

    make_option(c("-n", "--show-numbers"), default=FALSE, action="store_true",
                metavar="FALSE", type="logical",
                help="in violin with significance, show all numbers instead of stars"),

    make_option(c("-f", "--fluorescence"), default="zygosity",
                metavar="zygosity", type="character",
                help="use this to split out the fluorescence column from the group by"),

    make_option(c("-x", "--xlabel"), default="Comp-GFP-A :: mNeonGreen",
                metavar="Comp-GFP-A :: mNeonGreen", type="character",
                help="x-axis label for histograms"),

    make_option(c("-l", "--height"), default=500,
                metavar="500", type="integer",
                help="height in px"),

    make_option(c("-w", "--width"), default=2000,
                metavar="2000", type="integer",
                help="width in px, max width is 200000"),

    make_option(c("-t", "--troubleshooting"), default=FALSE, action="store_true",
                metavar="FALSE", type="logical",
                help="enable if troubleshooting to prevent overwriting your files")
)
opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)
troubleshooting = opt[['troubleshooting']]
output_dir <- file.path(wd, opt[['output-dir']])
troubleshooting_dir = file.path(output_dir, 'troubleshooting')

if (!(opt[['stat']] %in% c('fishers_lsd', 't_test', 'tukey', 'bonferroni'))) {
    stop("Choose from stat= 'fishers_lsd', 't_test', 'tukey', or 'bonferroni'")
}

# args
metadata_cols <- unlist(strsplit(opt[['group-by']], ','))

# Start Log
start_time = Sys.time()
log <- log_open(paste0("analyze_mfi-",
    strftime(start_time, format="%Y%m%d_%H%M%S"), '.log'))
log_print(paste('Script started at:', start_time))


# ----------------------------------------------------------------------
# Read Data

log_print(paste(Sys.time(), 'Reading data...'))


# import mfi
df <- import_flowjo_export(
    file.path(wd, opt[['input-dir']]),
    metric_name=opt[['metric']],
    include_initial_gates=FALSE
)
if (is.null(df)) {
    stop(paste('No files found in', file.path(wd, opt[['input-dir']])))
}


# left join sdev
sdev_df <- import_flowjo_export(file.path(wd, opt[['sdev-dir']]), metric_name='sdev')
if (!is.null(sdev_df)) {
    log_print(paste(Sys.time(), 'Merging sdev data...'))
    df <- merge(df, sdev_df,
        by=c("fcs_name", "gate", "cell_type"),
        all.x=TRUE, all.y=FALSE, suffixes=c('', '_'))
    df[(is.na(df[['sdev']])), 'sdev'] <- 0  # fillna
}


# left join counts
counts_df <- import_flowjo_export(
    file.path(wd, opt[['counts-dir']]),
    metric_name='num_cells',
    include_initial_gates=TRUE
)
if (!is.null(counts_df)) {
    log_print(paste(Sys.time(), 'Merging counts data...'))

    counts_cols <- c("fcs_name", "gate", "cell_type",  "num_cells")  # don't need pct_cells
    counts_df <- counts_df[, counts_cols]

    ss_df <- import_flowjo_export(
        file.path(wd, opt[['counts-dir']]),
        metric_name='num_cells',
        include_initial_gates=FALSE
    )
    ss_df <- ss_df[
        ((ss_df[['gate']]=='Cells/Single Cells/Single Cells') |
         (ss_df[['gate']]=='Cells/Single Cells/Single Cells/Live Cells')),
        counts_cols
    ]
    counts_df <- rbind(counts_df, ss_df)

    df <- merge(df, counts_df,
        by=c("fcs_name", "gate", "cell_type"),
        all.x=TRUE, all.y=FALSE, suffixes=c('', '_'))
}


# inner join flow metadata
flow_metadata <- import_flow_metadata(file.path(wd, opt[['metadata-dir']]))
df <- merge(df, flow_metadata,
    by="fcs_name", all.x=FALSE, all.y=FALSE, suffixes=c('', '_'))
df <- df[(df[['is_unstained']]==FALSE), ]  # drop unstained cells
if (length(metadata_cols) > 1) {
    df[['group_name']] <- apply( df[ , metadata_cols ] , 1 , paste , collapse = ", " )
    df[['subgroup_name']] <- apply(
        df[ , items_in_a_not_b(metadata_cols, opt[['fluorescence']]) , drop=FALSE],
        1 , paste , collapse = ", " )
} else {
    df[['group_name']] <- df[[metadata_cols]]
}


# left join mouse data
mouse_db <- append_many_csv(
    file.path(wd, opt[['ref-dir']]), recursive=TRUE, include_filepath=FALSE)
if (!is.null(mouse_db)) {
    log_print(paste(Sys.time(), 'Merging mouse data...'))
    df <- merge(df, mouse_db[, items_in_a_not_b(colnames(mouse_db), mouse_db_ignore)],
        by="mouse_id", all.x=TRUE, all.y=FALSE, suffixes=c('', '_'))
    df[['weeks_old']] <- round(df[['age']]/7, 1)
}


# filter and sort
df <- df[order(df[['cell_type']], df[['organ']], df[['group_name']]), ]  # sort rows


# ----------------------------------------------------------------------
# Data Overview

organs <- sort(unique(df[['organ']]))
log_print(paste(Sys.time(), 'Organs included...', paste(organs, collapse = ', ')))

for (organ in sort(organs)) {

    log_print(paste(Sys.time(), 'Plotting', organ, 'overview...'))

    # ------------------------------------------------------------
    # Data

    tmp <- sort_groups_by_metric(
        df[(df[['organ']]==organ),
            unique(intersect(
                c('organ', 'genotype', 'treatment',
                   'group_name', metadata_cols, 'cell_type',
                   'Cells/Single Cells/Single Cells/Live Cells',
                   'num_cells', opt[['metric']], 'viable_cells_conc',
                   'total_vol', 'total_viable_cells', 'abs_count',
                   'mouse_id', 'sex', 'weeks_old', 'fcs_name'),
                colnames(df)
            ))],
        x='cell_type',
        y=opt[['metric']],
        groups=c('group_name')
    )

    if (!troubleshooting) {
        dirpath <- file.path(wd, opt[['output-dir']], 'data',
            gsub(',', '_', opt[['group-by']]), opt[['metric']])
        if (!dir.exists(dirpath)) { dir.create(dirpath, recursive=TRUE) }
        filepath = file.path(dirpath, paste0(opt[['metric']], '-', organ, '.csv'))
        write.table(tmp, file = filepath, row.names = FALSE, sep = ',')
    }

    # ------------------------------------------------------------
    # Plot

    fig <- plot_violin(
        tmp,
        x='cell_type', y=opt[['metric']], group_by='group_name',
        ylabel='MFI', title=organ,
        ymin=0,
        hover_data=unique(intersect(
                c('mouse_id', 'group_name', metadata_cols,
                  'sex', 'treatment', 'weeks_old', 
                  'Cells', 'num_cells', 'fcs_name'),
                colnames(df)
        ))
    )

    if (!troubleshooting) {
        save_fig(
            fig=fig,
            height=opt[['height']], width=opt[['width']],
            dirpath=file.path(wd, opt[['output-dir']], 'figures',
                gsub(',', '_', opt[['group-by']]), opt[['metric']], 'interactive'),
            filename=paste(organ, opt[['metric']], sep='-'),
            save_html=TRUE
        )
    }

}


# ----------------------------------------------------------------------
# Total Variation Distance

if (!is.null(sdev_df)) {

    # ------------------------------------------------------------
    # Data

    log_print(paste(Sys.time(), 'Computing total variation distances...'))

    group_cols <- c('organ', 'gate', 'cell_type', metadata_cols)
    val_cols <- c('gmfi', 'sdev')
    tvd_df <- df[
        ((df[opt[['fluorescence']]]!='WT') &
         (df[['sdev']] > 0) & (df[['num_cells']] > 5)),
        c('mouse_id', 'fcs_name', group_cols, 'num_cells', val_cols)]

    # left join WT mean MFI
    wt_df <- group_by_agg(
        df[((df[opt[['fluorescence']]]=='WT') &
            (df[['sdev']] > 0) & (df[['num_cells']] > 5)),
           c(group_cols, val_cols)],
        group_cols, val_cols, mean
    )
    wt_df <- rename_columns(wt_df, c('gmfi'='wt_gmfi', 'sdev'='wt_sdev'))
    
    tvd_df <- merge(
        tvd_df, wt_df,
        by=items_in_a_not_b(group_cols, opt[['fluorescence']]),
        all.x=TRUE, all.y=FALSE, suffixes=c('', '_'))[,
        c('mouse_id', 'fcs_name', group_cols,
          'num_cells', val_cols, 'wt_gmfi', 'wt_sdev')
    ]
    tvd_df[['group_name']] <- apply(
        tvd_df[ , metadata_cols ] , 1 , paste , collapse = ", " )

    tvd_df[['tvd']] <- mapply(
        compute_normal_tvd,
        mean1 = tvd_df[['gmfi']],
        sd1   = tvd_df[['sdev']],
        mean2 = tvd_df[['wt_gmfi']],
        sd2   = tvd_df[['wt_sdev']],
        log_transform=TRUE
    )

    if (!troubleshooting) {
        dirpath <- file.path(wd, opt[['output-dir']], 'data',
            gsub(',', '_', opt[['group-by']]), opt[['metric']])
        # if (!dir.exists(dirpath)) { dir.create(dirpath, recursive=TRUE) }
        write.table(tvd_df, file=file.path(dirpath, paste0('gmfi-tvd', '.csv')),
            row.names=FALSE, sep=',' )
    }

    # ------------------------------------------------------------
    # Plot

    organs <- sort(unique(tvd_df[['organ']]))
    for (organ in sort(organs)) {

        log_print(paste(Sys.time(), 'Plotting', organ, 'tvd...'))

        fig <- plot_violin(
            tvd_df[(tvd_df[['organ']]==organ), ],
            x='cell_type', y='tvd', group_by='group_name',
            ylabel='TVD', title=organ,
            ymin=0, ymax=1,
            hover_data=unique(intersect(
                    c('mouse_id', 'group_name', metadata_cols,
                      'sex', 'treatment', 'weeks_old', 
                      'Cells', 'num_cells', 'fcs_name'),
                    colnames(df)
            ))
        )

        if (!troubleshooting) {
            dirpath <- file.path(wd, opt[['output-dir']], 'figures',
                gsub(',', '_', opt[['group-by']]), opt[['metric']], 'interactive')
            # if (!dir.exists(dirpath)) { dir.create(dirpath, recursive=TRUE) }
            save_fig(
                fig=fig,
                height=opt[['height']], width=opt[['width']],
                dirpath=dirpath,
                filename=paste(organ, 'tvd', sep='-'),
                save_html=TRUE
            )
        }
    }
}


# ----------------------------------------------------------------------
# MFI Multiple Comparisons

log_print(paste(Sys.time(), paste0('Computing ', opt[['stat']], '...')))

if (opt[['stat']]=='t_test') {
    pval_tbl <- apply_unpaired_t_test(
        df,
        index_cols=c('organ', 'cell_type'),
        group_name='group_name',
        metric=opt[['metric']]
    )
} else {
    pval_tbl <- apply_multiple_comparisons(
        df,
        index_cols=c('organ', 'cell_type'),
        group_name='group_name',
        metric=opt[['metric']],
        correction=opt[['stat']]
    )
}

log_print(paste(Sys.time(), paste(
    nrow(pval_tbl), 'populations found across',
    length(organs), 'organs') )
)

if (!troubleshooting) {
    tmp <- pval_tbl
    group_names <- sort(unique( df[['group_name']] ))
    for (col in group_names) {
        tmp[[col]] <- mapply(toJSON, tmp[[col]])
    }

    dirpath <- file.path(wd, opt[['output-dir']], 'data',
        gsub(',', '_', opt[['group-by']]), opt[['metric']])
    # if (!dir.exists(dirpath)) { dir.create(dirpath, recursive=TRUE) }
    filepath = file.path(dirpath, paste0('gmfi-pvals-', opt[['stat']], '.csv'))
    write.table(tmp, file = filepath, row.names = FALSE,  sep = ',' )
}


log_print(paste(Sys.time(), paste('Plotting mulitple comparisons using', opt[['stat']])))
if (opt[['plot-histograms']] && !is.null(sdev_df)) {
    log_print(paste(Sys.time(), 'Plotting fluorescence histograms...'))
}

pbar <- progress_bar$new(total = nrow(pval_tbl))
for (idx in 1:nrow(pval_tbl)) {

    # subset
    organ <- pval_tbl[idx, 'organ'][[1]]
    cell_type <- pval_tbl[idx, 'cell_type'][[1]]
    df_subset <- df[
        (df[['organ']] == organ) & (df[['cell_type']] == cell_type),
        unique(intersect(
                c('mouse_id', 'organ', 'cell_type', 'group_name',
                  'subgroup_name', opt[['fluorescence']],
                   opt[['metric']], 'sdev', 'num_cells', 'abs_count'),
                colnames(df)
        ))
    ]

    # ----------------------------------------------------------------------
    # MFI Violin Plot

    custom_group_order <- move_list_items_to_front(
        unique(df_subset[['group_name']]),
        custom_group_order
    )
    n_groups <- length(unique(custom_group_order))

    withCallingHandlers({
        fig <- plot_multiple_comparisons(
            df_subset[, c("organ", "cell_type", "group_name", opt[['metric']])],
            x='group_name', y=opt[['metric']],
            ymin=min( df_subset[, opt[['metric']]]*1.1, 0 ),
            ylabel=opt[['metric']],
            title=paste(toupper(organ), cell_type),
            test=opt[['stat']],
            show_numbers=opt[['show-numbers']],
            custom_group_order=custom_group_order  # manual input
        )
    }, warning = function(w) {
        if ( any(grepl("containing non-finite values", w),
                 grepl("outside the scale range", w)) ) {
            invokeRestart("muffleWarning")
        }
    })

    # save
    if (!troubleshooting) {
        dirpath <- file.path(wd, opt[['output-dir']], 'figures',
            gsub(',', '_', opt[['group-by']]), opt[['metric']], organ,
            opt[['stat']]
        )
        if (!dir.exists(dirpath)) { dir.create(dirpath, recursive=TRUE) }
        
        filepath = file.path(dirpath,
            paste0(gsub(' ', '', tolower(organ)), '-',
                gsub(' ', '_', tolower(cell_type)), '.svg' )
        )
        withCallingHandlers({
            ggsave(
                filepath,
                plot=fig,
                height=5000, width=5000, dpi=500, units='px',
                scaling=(if (n_groups==2) {1} else 2)
            )
        }, warning = function(w) {
            if ( any(grepl("containing non-finite values", w),
                     grepl("outside the scale range", w),
                     grepl("fewer than two data points", w),
                     grepl("argument is not numeric or logical: returning NA", w)) ) {
                invokeRestart("muffleWarning")
            }
        })
    }

    # ----------------------------------------------------------------------
    # Histograms

    if (opt[['plot-histograms']] && !is.null(sdev_df)) {

        subgroups <- unique(df_subset[['subgroup_name']])
        for (subgroup in subgroups) {

            df_subsubset <- df_subset[
                (df_subset['subgroup_name'] == subgroup),
                unique(intersect(
                        c('organ', 'mouse_id', 'group_name', 'subgroup_name', 'zygosity',
                          'num_cells', 'gmfi', 'sdev'),
                        colnames(df_subset)
                ))
            ]
            df_subsubset <- df_subsubset [!duplicated(df_subsubset[c('mouse_id', 'group_name')]), ]
            df_subsubset[['color']] <- multiple_replacement(
                df_subsubset[['zygosity']],
                c('homo'='#ACF53B', 'het'='#ACF53B', 'hemi'='#ACF53B', 'WT'='#B0B0B0')
            )

            sim_fluors <- Map(
                generate_lognormal_data,
                n = min(df_subsubset[['num_cells']], 1000, na.rm=TRUE),
                mean = df_subsubset[['gmfi']],
                sd = df_subsubset[['sdev']],
                group_name = apply(
                    df_subsubset[ , c('group_name', 'mouse_id') ],
                    1 , paste , collapse = ", " )
            )
            sim_fluor <- do.call(rbind, sim_fluors)

            fig <- plot_modal_histograms(sim_fluor,
                xlabel = opt[['xlabel']],
                ylabel = "Normalized to Mode",
                title = paste(toupper(organ), cell_type, subgroup),
                spar = 0.4,
                colors = df_subsubset[['color']],  # autogenerate this
                x_ticks = c(0, 250, 500, 1000, 2500, 5000, 10000),
                show_bins=FALSE
            )

            # save
            if (!troubleshooting) {
                dirpath <- file.path(wd, opt[['output-dir']], 'figures',
                    gsub(',', '_', opt[['group-by']]), opt[['metric']],
                    organ, 'histograms'
                )
                if (!dir.exists(dirpath)) { dir.create(dirpath, recursive=TRUE) }
                
                filepath = file.path(dirpath,
                    paste0(gsub(' ', '', tolower(organ)), '-',
                        gsub(' ', '_', tolower(cell_type)), '-',
                        gsub(', ', '_', subgroup), '.png' )
                )
                ggsave(
                    filepath,
                    plot=fig, bg='white',
                    height=1500, width=3000, dpi=200, units='px', scaling=2
                )
            }
        }
    }

    pbar$tick()
}


end_time = Sys.time()
log_print(paste('Script ended at:', Sys.time()))
log_print(paste("Script completed in:", difftime(end_time, start_time)))
log_close()

File: python/analyze_mfi.py
#!/usr/bin/env python3
"""
Graphs MFI (Mean Fluorescence Intensity) for each cell population.
Supports geometric mean (recommended), median, or arithmetic mean.
"""

import os
import sys
import argparse
import logging
from datetime import datetime
from tqdm import tqdm

import pandas as pd
import numpy as np

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from functions.preprocessing import sort_groups_by_metric
from functions.file_io import import_flowjo_export, import_flow_metadata
from tools.file_io import append_many_csv
from tools.df_tools import group_by_agg, rename_columns
from tools.list_tools import items_in_a_not_b, move_list_items_to_front
from tools.math import (apply_unpaired_t_test, apply_multiple_comparisons,
                        generate_lognormal_data, compute_normal_tvd)
from tools.plotting import (save_fig, plot_violin, plot_multiple_comparisons,
                            plot_modal_histograms)
from config.flow import mouse_db_ignore
from config.user_input import custom_group_order


def main():
    parser = argparse.ArgumentParser(description='Analyze flow cytometry MFI')
    parser.add_argument('-i', '--input-dir', default='data/flow-mfi',
                        help='Input MFI directory')
    parser.add_argument('-o', '--output-dir', default='hfc-output',
                        help='Output directory')
    parser.add_argument('-d', '--sdev-dir', default='data/flow-sdev',
                        help='Standard deviations directory')
    parser.add_argument('-c', '--counts-dir', default='data/flow-counts',
                        help='Counts directory')
    parser.add_argument('-m', '--metadata-dir', default='data/flow-metadata',
                        help='Metadata directory')
    parser.add_argument('-r', '--ref-dir', default='data/mice',
                        help='Mouse reference directory')
    parser.add_argument('-g', '--group-by', default='sex,treatment,zygosity',
                        help='Grouping columns')
    parser.add_argument('-p', '--plot-histograms', action='store_true',
                        help='Generate histograms')
    parser.add_argument('-y', '--metric', default='gmfi',
                        help='Metric name (gmfi, mode)')
    parser.add_argument('-s', '--stat', default='fishers_lsd',
                        choices=['fishers_lsd', 't_test', 'tukey', 'bonferroni'])
    parser.add_argument('-n', '--show-numbers', action='store_true')
    parser.add_argument('-f', '--fluorescence', default='zygosity',
                        help='Fluorescence grouping column')
    parser.add_argument('-x', '--xlabel', default='Comp-GFP-A :: mNeonGreen',
                        help='X-axis label for histograms')
    parser.add_argument('-l', '--height', type=int, default=500)
    parser.add_argument('-w', '--width', type=int, default=2000)
    parser.add_argument('-t', '--troubleshooting', action='store_true')

    args = parser.parse_args()

    # Setup
    wd = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    output_dir = os.path.join(wd, args.output_dir)
    metadata_cols = args.group_by.split(',')

    # Logging
    start_time = datetime.now()
    logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(message)s')
    logging.info(f'Script started at: {start_time}')

    # Read MFI data
    logging.info('Reading data...')
    df = import_flowjo_export(
        os.path.join(wd, args.input_dir),
        metric_name=args.metric, include_initial_gates=False
    )

    if df is None:
        raise ValueError(f'No files found in {os.path.join(wd, args.input_dir)}')

    # Join sdev
    sdev_df = import_flowjo_export(
        os.path.join(wd, args.sdev_dir), metric_name='sdev'
    )
    if sdev_df is not None:
        logging.info('Merging sdev data...')
        df = df.merge(sdev_df, on=['fcs_name', 'gate', 'cell_type'],
                      how='left', suffixes=('', '_'))
        df['sdev'] = df['sdev'].fillna(0)

    # Join counts
    counts_df = import_flowjo_export(
        os.path.join(wd, args.counts_dir),
        metric_name='num_cells', include_initial_gates=True
    )
    if counts_df is not None:
        logging.info('Merging counts data...')
        df = df.merge(counts_df[['fcs_name', 'gate', 'cell_type', 'num_cells']],
                      on=['fcs_name', 'gate', 'cell_type'],
                      how='left', suffixes=('', '_'))

    # Join metadata
    flow_metadata = import_flow_metadata(os.path.join(wd, args.metadata_dir))
    df = df.merge(flow_metadata, on='fcs_name', how='inner', suffixes=('', '_'))

    if 'is_unstained' in df.columns:
        df = df[df['is_unstained'] == False]

    # Create group names
    if len(metadata_cols) > 1:
        df['group_name'] = df[metadata_cols].apply(
            lambda x: ', '.join(x.astype(str)), axis=1
        )
        subgroup_cols = [c for c in metadata_cols if c != args.fluorescence]
        df['subgroup_name'] = df[subgroup_cols].apply(
            lambda x: ', '.join(x.astype(str)), axis=1
        )
    else:
        df['group_name'] = df[metadata_cols[0]]

    # Join mouse data
    mouse_db = append_many_csv(os.path.join(wd, args.ref_dir),
                               recursive=True, include_filepath=False)
    if mouse_db is not None:
        logging.info('Merging mouse data...')
        mouse_cols = items_in_a_not_b(list(mouse_db.columns), mouse_db_ignore)
        df = df.merge(mouse_db[mouse_cols], on='mouse_id',
                      how='left', suffixes=('', '_'))
        if 'age' in df.columns:
            df['weeks_old'] = (df['age'] / 7).round(1)

    # Sort
    df = df.sort_values(['cell_type', 'organ', 'group_name'])

    # Process each organ
    organs = sorted(df['organ'].unique())
    logging.info(f'Organs included: {", ".join(organs)}')

    for organ in organs:
        logging.info(f'Plotting {organ} overview...')

        organ_df = df[df['organ'] == organ].copy()
        organ_df = sort_groups_by_metric(
            organ_df, x='cell_type', y=args.metric, groups=['group_name']
        )

        if not args.troubleshooting:
            dirpath = os.path.join(output_dir, 'data',
                                   args.group_by.replace(',', '_'), args.metric)
            os.makedirs(dirpath, exist_ok=True)
            organ_df.to_csv(os.path.join(dirpath, f'{args.metric}-{organ}.csv'),
                            index=False)

            fig = plot_violin(
                organ_df, x='cell_type', y=args.metric, group_by='group_name',
                ylabel='MFI', title=organ, ymin=0
            )
            save_fig(
                fig, height=args.height, width=args.width,
                dirpath=os.path.join(output_dir, 'figures',
                                     args.group_by.replace(',', '_'),
                                     args.metric, 'interactive'),
                filename=f'{organ}-{args.metric}',
                save_html=True
            )

    # Compute statistics
    logging.info(f'Computing {args.stat}...')

    if args.stat == 't_test':
        pval_tbl = apply_unpaired_t_test(
            df, index_cols=['organ', 'cell_type'],
            group_name='group_name', metric=args.metric
        )
    else:
        pval_tbl = apply_multiple_comparisons(
            df, index_cols=['organ', 'cell_type'],
            group_name='group_name', metric=args.metric,
            correction=args.stat
        )

    if pval_tbl is not None:
        logging.info(f'{len(pval_tbl)} populations found across {len(organs)} organs')

        if not args.troubleshooting:
            dirpath = os.path.join(output_dir, 'data',
                                   args.group_by.replace(',', '_'), args.metric)
            pval_tbl.to_csv(os.path.join(dirpath, f'gmfi-pvals-{args.stat}.csv'),
                            index=False)

        # Plot multiple comparisons
        logging.info(f'Plotting multiple comparisons using {args.stat}')

        for idx in tqdm(range(len(pval_tbl)), desc='MFI comparisons'):
            organ = pval_tbl.iloc[idx]['organ']
            cell_type = pval_tbl.iloc[idx]['cell_type']

            df_subset = df[(df['organ'] == organ) & (df['cell_type'] == cell_type)]

            if len(df_subset) == 0:
                continue

            local_group_order = move_list_items_to_front(
                df_subset['group_name'].unique().tolist(),
                custom_group_order
            )

            try:
                fig = plot_multiple_comparisons(
                    df_subset[['organ', 'cell_type', 'group_name', args.metric]],
                    x='group_name', y=args.metric,
                    ylabel=args.metric,
                    title=f'{organ.upper()} {cell_type}',
                    test=args.stat,
                    show_numbers=args.show_numbers,
                    custom_group_order=local_group_order
                )

                if not args.troubleshooting:
                    dirpath = os.path.join(output_dir, 'figures',
                                           args.group_by.replace(',', '_'),
                                           args.metric, organ, args.stat)
                    os.makedirs(dirpath, exist_ok=True)

                    filename = f"{organ.lower().replace(' ', '')}-{cell_type.lower().replace(' ', '_')}.svg"
                    fig.savefig(os.path.join(dirpath, filename),
                                dpi=500, bbox_inches='tight')
                    plt.close(fig)
            except Exception as e:
                logging.warning(f"Error plotting {organ} {cell_type}: {e}")

    end_time = datetime.now()
    logging.info(f'Script ended at: {end_time}')
    logging.info(f'Script completed in: {end_time - start_time}')


if __name__ == '__main__':
    import matplotlib.pyplot as plt
    main()
